### PR TITLE
docs: add first cut adapter implementation API doc

### LINKED ISSUE #304: Document the new adapter API
The adapter API added in #248 should be documented somewhere to to make the requirements clear for adapter implementations.

### PR DESCRIPTION
**Pull request by demmer**
First cut at describing the adapter API.

Resolves #304.

### CODE REVIEW

**Comment by demmer** [comment]
@rlgomes @dmehra @mstemm take a look.

**Comment by mstemm** [comment]
I think adding a short overview section that describes at a high level the necessary steps in creating an adapter would be useful, those steps being:

1. Create a javascript module that is require()d when you run a program using the adapter.
2. In that module, implement an init() function to define the read/write capabilities of your adapter.
3. Implement ES6 classes deriving from AdapterRead/AdapterWrite that perform the work of reading from/writing to your backend.
4. The juttle runtime will instantiate your read/write classes every time a program is run.

File: docs/adapters/adapter_api.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+The adapter should reference the runtime using this API only and should not use a node.js `require()` call to pull in portions of the juttle compiler or runtime.
+
+## Initialization
+
+Each adapter is implemented as a node.js module that exports a single initializer function.
+
+### initialize(config)
+
+Initialize the adapter. Paramters include:
```
Comment (rlgomes) [inline docs/adapters/adapter_api.md]: `s/Paramters/Parameters/`

File: docs/adapters/adapter_api.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+### start()
+
+Called when the program activates.
+
+### read(from, to, limit, state)
+
+Read a set of points. Parameters indicate:
+
+* from: starting time from which to read
```
Comment (rlgomes) [inline docs/adapters/adapter_api.md]: Should we explain how to use JuttleMoments since the from/to are JuttleMoments.

**Comment by mstemm** [comment]
Also would this be the right doc to include version support and the juttleAdapterAPI property in package.json?

File: docs/adapters/adapter_api.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
+# Juttle Adapter API
+
+This document describes the low level API used to implement juttle adapters.
```
Comment (dmehra) [inline docs/adapters/adapter_api.md]: Does it make sense to link from here to https://github.com/juttle/juttle-adapter-template as an example of implementation that uses the API? And perhaps to the gmail adapter if it's still meant to be a "teaching adapter"?

File: mkdocs.yml [hunk 4]
```diff
@@ -90,7 +90,8 @@ pages:
    - '____ CDN Source': 'adapters/stochastic_cdn.md'
    - '____ Logs Source': 'adapters/stochastic_logs.md'
    - '____ Search Cluster Source': 'adapters/stochastic_search_cluster.md'
-   
+   - '____ Adapter Implementation API': 'adapters/adapter_api.md'
```
Comment (dmehra) [inline mkdocs.yml]: This would hide the doc from the left pane. How should it be findable by developers?

**Comment by demmer** [comment]
@mstemm I'm going to merge this as-is since it's a reasonable start and is better than nothing.

We should merge in the high level overview portions of the gmail adapter implementation guide as a separate PR.

**Pull request merged**